core/src/main/java/com/scalar/db/transaction/consensuscommit/RecoveryHandler.java

```diff
@@ -4,53 +4,167 @@
     if (state.isPresent()) {
       if (state.get().getState().equals(TransactionState.COMMITTED)) {
-        rollforwardRecord(selection, result);
+        rollforwardRecord(selection, latestResult.get());
+        return getAfterImageResult(latestResult.get(), metadata);
       } else {
-        rollbackRecord(selection, result);
+        rollbackRecord(selection, latestResult.get());
+        return getBeforeImageResult(latestResult.get(), metadata);
+      }
+    } else {
+      abortIfExpired(selection, latestResult.get());
+      return getBeforeImageResult(latestResult.get(), metadata);
+    }
+  }
+
+  private Optional<TransactionResult> getLatestResult(Selection selection, TransactionResult result)
+      throws RecoveryException {
+    try {
+      Get get =
+          new Get(selection.getPartitionKey(), getClusteringKey(selection, result).orElse(null))
+              .withConsistency(Consistency.LINEARIZABLE)
+              .forNamespace(selection.forNamespace().get())
+              .forTable(selection.forTable().get());
+      return storage.get(get).map(TransactionResult::new);
+    } catch (ExecutionException e) {
+      throw new RecoveryException("getting the latest result failed", e);
+    }
+  }
+
+  private Optional<Key> getClusteringKey(Selection selection, TransactionResult result) {
+    if (selection instanceof Scan) {
+      return result.getClusteringKey();
+    } else {
+      return selection.getClusteringKey();
+    }
+  }
+
+  private Optional<Coordinator.State> getState(String id) throws RecoveryException {
+    try {
+      return coordinator.getState(id);
+    } catch (CoordinatorException e) {
+      throw new RecoveryException("getting the coordinator state failed", e);
+    }
+  }
+
+  private TransactionalTableMetadata getTableMetadata(Selection selection)
+      throws RecoveryException {
+    try {
+      return tableMetadataManager.getTransactionalTableMetadata(selection);
+    } catch (ExecutionException e) {
+      throw new RecoveryException("getting the table metadata failed", e);
+    }
+  }
+
+  private Optional<TransactionResult> getAfterImageResult(
+      TransactionResult result, TransactionalTableMetadata metadata) {
+    if (result.getState().equals(TransactionState.PREPARED)) {
+      Map<String, Value<?>> values = new HashMap<>();
+      for (String afterImageColumnName : metadata.getAfterImageColumnNames()) {
+        assert result.getValue(afterImageColumnName).isPresent();
+        if (afterImageColumnName.equals(Attribute.STATE)) {
+          // change the state to COMMITTED
+          values.put(Attribute.STATE, Attribute.toStateValue(TransactionState.COMMITTED));
+        } else {
+          values.put(afterImageColumnName, result.getValue(afterImageColumnName).get());
+        }
+      }
+      return Optional.of(
+          new TransactionResult(new ResultImpl(values, metadata.getTableMetadata())));
+    } else { // TransactionState.DELETED
+      return Optional.empty();
+    }
+  }
```

Comment:
Note that this after image result returned by `getAfterImageResult()` doesn't have the `committed_at` column.